### PR TITLE
apiv2 - add first backend implementation of topAuthors queries

### PR DESCRIPTION
**Pull request by morucci**
Related to #468

### CODE REVIEW

File: haskell/src/Monocle/Backend/Test.hs [hunk 1]
```diff
@@ -301,16 +312,38 @@ testReposSummary = withTenant doTest
...
+      assertEqual' "Check getMostActiveAuthorByChangeCreated count" [Q.TermResult {term = "eve", count = 4}] results
+      results' <- runQueryM defaultQuery Q.getMostActiveAuthorByChangeMerged
+      assertEqual' "Check getMostActiveAuthorByChangeMerged count" [Q.TermResult {term = "eve", count = 2}] results'
+      results'' <- runQueryM defaultQuery Q.getMostActiveAuthorByChangeReviewed
+      assertEqual'
+        "Check getMostActiveAuthorByChangeReviewed count"
+        [ Q.TermResult {term = "alice", count = 2},
+          Q.TermResult {term = "bob", count = 2}
+        ]
+        results''
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Test.hs:338]: You can create longer QueryM using do notation:

```suggestion
      runQueryM defaultQuery $ do
        results <- Q.getMostActiveAuthorByChangeCreated
        assertEqual' "Check getMostActiveAuthorByChangeCreated count" [Q.TermResult {term = "eve", count = 4}] results
        results' <- Q.getMostActiveAuthorByChangeMerged
        assertEqual' "Check getMostActiveAuthorByChangeMerged count" [Q.TermResult {term = "eve", count = 2}] results'
        results'' <- Q.getMostActiveAuthorByChangeReviewed
        assertEqual'
          "Check getMostActiveAuthorByChangeReviewed count"
          [ Q.TermResult {term = "alice", count = 2},
            Q.TermResult {term = "bob", count = 2}
          ]
          results''
```

Comment (morucci) [reply haskell/src/Monocle/Backend/Test.hs:338]: Ok I'll do, but to be honest I'm not sure it clarify the code or at least this way to use the do notation is too new for me.

File: haskell/src/Monocle/Backend/Test.hs [hunk 2]
```diff
@@ -301,16 +312,38 @@ testReposSummary = withTenant doTest
...
-              fromMaybe (error "nop") (readMaybe "2099-12-31 23:59:59 Z")
-            )
-          queryMinBoundsSet = False
-       in Q.Query {..}
+
+testTopAuthors :: Assertion
+testTopAuthors = withTenant doTest
+  where
+    doTest :: TenantM ()
+    doTest = do
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Test.hs:320]: Another solution would be to do (and remove the inlined runQueryM below):
```suggestion
    doTest :: TenantM ()
    doTest = runQueryM defaultQuery doTestQuery
    
    doTestQuery :: QueryM ()
    doTestQuery = do
```

File: haskell/src/Monocle/Backend/Test.hs (hunk 1 above)
Comment (TristanCacqueray) [reply haskell/src/Monocle/Backend/Test.hs:338]: Using such do block is relatively common, there is even an extension to make it more easy by removing the `$`: https://typeclasses.com/ghc/block-arguments

File: haskell/src/Monocle/Backend/Test.hs (hunk 2 above)
Comment (TristanCacqueray) [reply haskell/src/Monocle/Backend/Test.hs:320]: Or using BlockArgument
```suggestion
    doTest :: TenantM ()
    doTest = runQueryM defaultQuery do
```

File: haskell/src/Monocle/Backend/Test.hs [hunk 3]
```diff
@@ -384,12 +464,13 @@ indexScenario xs = sequence_ $ indexDoc <$> xs
       SChange d -> I.indexChanges [d]
       SCreation d -> I.indexEvents [d]
       SReview d -> I.indexEvents [d]
+      SComment d -> I.indexEvents [d]
       SMerge d -> I.indexEvents [d]
 
 -- | 'nominalMerge' is the most simple scenario
 -- >>> let project = SProject "openstack/nova" [Author "alice" "a", Author "bob" "b"] [Author "eve" "e"]
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Test.hs:471]: It seems like you need to add the commenter list here.

File: haskell/src/Monocle/Backend/Test.hs [hunk 4]
```diff
@@ -409,28 +490,33 @@ nominalMerge SProject {..} changeId start duration = evalRand scenario stdGen
...
               }
 
       -- The change creation
       author <- randomAuthor contributors
-      let create = mkEvent 0 "ChangeCreatedEvent" author
+      let create = mkEvent 0 ChangeCreated author author
           change = mkChange 0 author
 
+      -- The comment
+      commenter <- randomAuthor commenters
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Test.hs:510]: Or perhaps commenters could be (maintainers <> contributors)

**Review by TristanCacqueray — APPROVED** [review]

**Comment by morucci** [comment]
This one is good to go. I'll move forward the server API in another PR + UI integration.

File: haskell/src/Monocle/Backend/Queries.hs [hunk 5]
```diff
@@ -312,23 +317,27 @@ getTermsAgg query onTerm = do
...
-
-getRepos :: QueryM [TermResult]
-getRepos = do
+getDocTypeTopCountByField :: Text -> Text -> QueryFlavor -> QueryM [TermResult]
+getDocTypeTopCountByField doctype attr qflavor = do
   -- Prepare the query
-  basequery <- toBHQueryWithFlavor (QueryFlavor OnAuthor CreatedAt) <$> getQuery
-  let query = mkAnd $ basequery <> getQueryE
+  basequery <- toBHQueryWithFlavor qflavor <$> getQuery
+  let query = mkAnd $ basequery <> [documentType doctype]
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Queries.hs:324]: nit: this pattern is so common, there is a generalized solution with [withFilter](https://github.com/change-metrics/monocle/blob/aefaff60f5c162f3d603a3f54d569ca77bcd7cce/haskell/src/Monocle/Servant/Env.hs#L122) and [getQueryBHWithFlavor](https://github.com/change-metrics/monocle/blob/aefaff60f5c162f3d603a3f54d569ca77bcd7cce/haskell/src/Monocle/Servant/Env.hs#L109). They can be used like so here:

```haskell
getDocTypeTopCountByField :: Text -> Text -> QueryFlavor -> QueryM [TermResult]
getDocTypeTopCountByField doctype attr qflavor = withFilter [documentType doctype] $ do
    query <- getQueryBHWithFlavor qflavor

    results <- liftTenantM (runTermAgg query)
    ...
```

**Review by TristanCacqueray — APPROVED** [review]
Well done!

**Pull request merged**